phoenix-core/src/main/java/org/apache/phoenix/coprocessor/UngroupedAggregateRegionScanner.java

```diff
@@ -0,0 +1,646 @@
+                            if (isDescRowKeyOrderUpgrade) {
+                                if (!descRowKeyOrderUpgrade(results, ptr, mutations)) {
+                                    continue;
+                                }
+                            } else if (buildLocalIndex) {
+                                buildLocalIndex(result, results, ptr, mutations);
+                            } else if (isDelete) {
+                                deleteRow(results, mutations);
+                            } else if (isUpsert) {
+                                upsert(result, ptr, mutations);
+                            } else if (deleteCF != null && deleteCQ != null) {
+                                deleteCForQ(result, results, mutations);
+                            }
+                            if (emptyCF != null) {
+                                /*
+                                 * If we've specified an emptyCF, then we need to insert an empty
+                                 * key value "retroactively" for any key value that is visible at
+                                 * the timestamp that the DDL was issued. Key values that are not
+                                 * visible at this timestamp will not ever be projected up to
+                                 * scans past this timestamp, so don't need to be considered.
+                                 * We insert one empty key value per row per timestamp.
+                                 */
+                                insertEmptyKeyValue(results, mutations);
+                            }
+                            if (ServerUtil.readyToCommit(mutations.size(), mutations.byteSize(), maxBatchSize, maxBatchSizeBytes)) {
+                                ungroupedAggregateRegionObserver.commit(region, mutations, indexUUID, blockingMemStoreSize, indexMaintainersPtr,
+                                        txState, targetHTable, useIndexProto, isPKChanging, clientVersionBytes);
+                                mutations.clear();
+                            }
+                            // Commit in batches based on UPSERT_BATCH_SIZE_BYTES_ATTRIB in config
+
+                            if (ServerUtil.readyToCommit(indexMutations.size(), indexMutations.byteSize(), maxBatchSize, maxBatchSizeBytes)) {
+                                setIndexAndTransactionProperties(indexMutations, indexUUID, indexMaintainersPtr, txState, clientVersionBytes, useIndexProto);
+                                ungroupedAggregateRegionObserver.commitBatch(region, indexMutations, blockingMemStoreSize);
+                                indexMutations.clear();
+                            }
+                            aggregators.aggregate(rowAggregators, result);
+                            hasAny = true;
+                        }
+                    } while (hasMore && (EnvironmentEdgeManager.currentTimeMillis() - startTime) < pageSizeInMs);
```

Comment:
This logic is not sufficient to ensure that only `pageSizeInMs` time is spent on the server-side operation. If say `pageSizeInMs=1000` and the last iteration starts at 990 ms but takes much longer than 10 ms, we can end up spending much more time than the configured page size duration. Is that fine?

If not, we can use something like `ExecutorService` with a timeout and do the entire `do while` logic inside a separate thread. This ensures that only `pageSizeInMs` time is ever spent on this. This might complicate things a bit however if there is work being done inside the loop which might be left in an incomplete state on thread interruption. Also, if we go with this approach, we don't guarantee even 1 iteration of the loop in spite of it being a `do while`.

Depends on how strict we want the paging duration limit to be. What do you think?

Reply:
The duration for a page within coproc will be a fraction of HBase client timeout value (1 second vs 2 minutes). As long as each iteration does not take too long, i.e., minutes, this Jira will be sufficient to eliminate the timeouts.  However,  an iteration of a region scanner can take minutes. This happens when the filter set on the scan is very selective or there are too many delete markers in the table regions. 6211 will address this remaining issue.

Reply:
Ok makes sense.